components/scream/src/physics/shoc/atmosphere_macrophysics.cpp

```diff
@@ -55,55 +115,269 @@ set_updated_group (const FieldGroup<Real>& group)
-void SHOCMacrophysics::initialize_impl (const util::TimeStamp& /* t0 */)
+void SHOCMacrophysics::initialize_impl (const util::TimeStamp& t0)
 {
-  // TODO: create the host mirrors once.
-  auto Q = m_shoc_groups_inout.at("TRACERS").m_bundle;
-  Q->sync_to_host();
+  m_current_ts = t0;
+
+  // Initialize all of the structures that are passed to shoc_main in run_impl.
+  // Note: Some variables in the structures are not stored in the field manager.  For these
+  //       variables a local view is constructed.
+
+  auto t        = m_shoc_fields_out["t"].get_reshaped_view<Spack**>();
+  auto alst     = m_shoc_fields_in["alst"].get_reshaped_view<const Spack**>();
+  auto zi       = m_shoc_fields_in["zi"].get_reshaped_view<const Spack**>();
+  auto zm       = m_shoc_fields_in["zm"].get_reshaped_view<const Spack**>();
+  auto pmid     = m_shoc_fields_in["pmid"].get_reshaped_view<const Spack**>();
+  auto pdel     = m_shoc_fields_in["pdel"].get_reshaped_view<const Spack**>();
+  auto omega    = m_shoc_fields_in["omega"].get_reshaped_view<const Spack**>();
+  auto shf      = m_shoc_fields_in["shf"].get_reshaped_view<const Pack1d*>();
+  auto cflx_k0  = m_shoc_fields_in["cflx_k0"].get_reshaped_view<const Pack1d*>();
+  auto wsx      = m_shoc_fields_in["wsx"].get_reshaped_view<const Pack1d*>();
+  auto wsy      = m_shoc_fields_in["wsy"].get_reshaped_view<const Pack1d*>();
+  auto shoc_ql  = m_shoc_fields_out["shoc_ql"].get_reshaped_view<Spack**>();
+  auto shoc_qv  = m_shoc_fields_out["shoc_qv"].get_reshaped_view<Spack**>();
+  auto tke      = m_shoc_fields_out["tke"].get_reshaped_view<Spack**>();
+  auto s        = m_shoc_fields_out["s"].get_reshaped_view<Spack**>();
+  auto u        = m_shoc_fields_out["u"].get_reshaped_view<Spack**>();
+  auto v        = m_shoc_fields_out["v"].get_reshaped_view<Spack**>();
+  auto Q        = m_shoc_fields_out["Q"].get_reshaped_view<Spack***>();
+
+  const int nlev_packs = ekat::npack<Spack>(m_num_levs);
+  const int nlevi_packs = ekat::npack<Spack>(m_num_levs+1);
+  const int num_tracer_packs = ekat::npack<Spack>(m_num_tracers);
+
+  view_1d wpthlp_sfc("wpthlp_sfc",m_num_cols),
+          wprtp_sfc("wprtp_sfc",m_num_cols),
+          upwp_sfc("upwp_sfc",m_num_cols),
+          vpwp_sfc("vpwp_sfc",m_num_cols);
+
+  view_2d rrho("rrho",m_num_cols,nlev_packs),
+          rrho_i("rrhoi",m_num_cols,nlevi_packs),
+          thv("thv",m_num_cols,nlev_packs),
+          dz("dz",m_num_cols,nlev_packs),
+          zt_grid("zt_grid",m_num_cols,nlev_packs),
+          zi_grid("zi_grid",m_num_cols,nlevi_packs),
+          wtracer_sfc("wtracer_sfc",m_num_cols,num_tracer_packs),
+          wm_zt("wm_zt",m_num_cols,nlev_packs),
+          exner("exner",m_num_cols,nlev_packs),
+          thlm("thlm",m_num_cols,nlev_packs),
+          qw("qw",m_num_cols,nlev_packs),
+          cloud_frac("cloud_frac",m_num_cols,nlev_packs);
+
+  view_3d tracers("tracers",m_num_cols,m_num_levs,num_tracer_packs);
```

Comment:
So, tbc, in the optimization follow-up PR, you won't allocate a 3D tracer array here?

Reply:
Yes, I'll basically just pass the Q array to shoc_main and the transpose will be handled internally.